src/czar/CzarFamilyMap.cc

```diff
@@ -0,0 +1,263 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "czar/CzarFamilyMap.h"
+
+// System headers
+#include <sstream>
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+// Qserv headers
+#include "qmeta/QMeta.h"
+#include "cconfig/CzarConfig.h"
+#include "czar/Czar.h"
+#include "czar/CzarRegistry.h"
+#include "qmeta/Exceptions.h"
+#include "util/Bug.h"
+#include "util/TimeUtils.h"
+
+using namespace std;
+
+namespace {
+LOG_LOGGER _log = LOG_GET("lsst.qserv.czar.CzarFamilyMap");
+}  // namespace
+
+namespace lsst::qserv::czar {
+
+CzarFamilyMap::Ptr CzarFamilyMap::create(std::shared_ptr<qmeta::QMeta> const& qmeta) {
+    // There's nothing the czar can do until with user queries until there's been at least
+    // one successful read of the database family tables, as the czar doesn't know where to find anything.
+    Ptr newPtr = nullptr;
+    while (newPtr == nullptr) {
+        try {
+            newPtr = Ptr(new CzarFamilyMap(qmeta));
+        } catch (ChunkMapException const& exc) {
+            LOGS(_log, LOG_LVL_WARN, "Could not create CzarFamilyMap, sleep and retry " << exc.what());
+        }
+        if (newPtr == nullptr) {
+            this_thread::sleep_for(10s);
+        }
+    }
+
+    return newPtr;
+}
+
+CzarFamilyMap::CzarFamilyMap(std::shared_ptr<qmeta::QMeta> const& qmeta) : _qmeta(qmeta) {
+    try {
+        auto mapsSet = _read();
+        if (!mapsSet) {
+            throw ChunkMapException(ERR_LOC, cName(__func__) + " maps were not set in constructor");
+        }
+    } catch (qmeta::QMetaError const& qExc) {
+        LOGS(_log, LOG_LVL_ERROR, cName(__func__) << " could not read DB " << qExc.what());
+        throw ChunkMapException(ERR_LOC, cName(__func__) + " constructor failed read " + qExc.what());
+    }
+}
+
+bool CzarFamilyMap::read() {
+    bool mapsSet = false;
+    try {
+        mapsSet = _read();
+    } catch (qmeta::QMetaError const& qExc) {
+        LOGS(_log, LOG_LVL_ERROR, cName(__func__) + " could not read DB " << qExc.what());
+    }
+    return mapsSet;
+}
+
+bool CzarFamilyMap::_read() {
+    LOGS(_log, LOG_LVL_TRACE, "CzarFamilyMap::_read() start");
+    // If replacing the map, this may take a bit of time, but it's probably
+    // better to wait for new maps if something changed.
+    std::lock_guard gLock(_familyMapMtx);
+    qmeta::QMetaChunkMap qChunkMap = _qmeta->getChunkMap(_lastUpdateTime);
+    if (_lastUpdateTime == qChunkMap.updateTime) {
+        // If "_lastUpdateTime == qChunkMap.updateTime", qChunkMap is empty.
```

Comment:
Is this statement really true? Asking this since I noticed the following assignment was made later in the current method after successfully reading the map from the database:
```
_lastUpdateTime = qChunkMap.updateTime;
```